Correcting issues with Marshmallow, relaxing NOT NULL constraint on m…

socketbox opened the pull request:
…odel for item and correcting duplicate home column for appuser

Summary of changes:

- moved marshmallow instance to pryce.database.dal
- for any `add_<model_obj>` method in the data access layer, returning the PK of the newly added instance (on success)
- fixed the NOT NULL constraint on Item description and removed an extraneous/duplicate reference to the home column in appuser
- added the beginning of a mock object creation class

jpcasey commented:
Thanks Casey.

Don't merge this in yet.  There are a couple conflicts based on some changes I needed to make last night when implementing the routes for Khoa to start testing things.  https://github.com/jpcasey/pryce/commit/974c2b99e89052e15a6a324dffabcc5d61eed2c9

socketbox commented:
> Added a couple comments on the conflicts.
> 
> I saw you removed the location table in a previous commit, and I know we spoke on that briefly the other day. But I do still think we should still have some location data in our db for each store. Could you add lat/long columns to the store table?

I was under the impression that we were eschewing lat/long in favor of `place_id`. The store model has a `place_id` field (a string) in lieu of a location FK or a lat/long pair. But I'm game for whatever you think is necessary. If we want lat/long we should decide on what format the coordinates take.

jpcasey commented:
> I was under the impression that we were eschewing lat/long in favor of `place_id`. The store model has a `place_id` field (a string) in lieu of a location FK or a lat/long pair. But I'm game for whatever you think is necessary. If we want lat/long we should decide on what format the coordinates take.

Yeah, I'm still for using place_id for lookups.

Just thinking to the future, it might be nice to still have access to those coordinates in our own db.  I'm not sure if we'd end up doing anything with them for this class, but I know Khoa was mentioning before about placing pins on a map for nearby reported prices.  Storing lat/long data in our table would allow us to implement something internally that handles that, without the need to make any expensive API calls.  For the shopping list functionality where we need to return stores with the lowest price, it could also allow us to potentially add a distance filter or something based on the user location, too.  Not a hard requirement but nice to have.  

I'm not sure about format.  Looking at the Places API calls lat/long are usually 6 digit precision? (e.g. "location": {
"lat": 40.4464055,
"lng": -80.183559
})

socketbox commented on pryce/database/dal/item.py:
```diff
@@ -13,21 +14,16 @@ def get_items(self):
     def get_item(self, code):
+        item = None
         try:
             item = Item.query.filter_by(code=code).one()
```
No, that was a mistake on my part that I just corrected (globally, across all the dal classes).

socketbox commented:
Merged your most recent commit to master into this branch. 

Followed your lead and created a document for the methods in the data access layer. 

I've updated the models and the migrations scripts. Store now has a `lat` column and a `lng` column. There's a Postgres POINT data type, but it's too close to Monday to mess around with figuring out how that interacts with SQLAlchemy (it looks like it needs another library).

Here are some questions that arose:

- It's debatable whether we should have a client-facing `add_item` endpoint, given that no item will be added absent a price, and we need to call `DALItem.add_item()` internally prior to adding a price. Regardless, the `DALItem.add_item()` method is taking an item instance, which is only feasible if we have created an item from JSON. Looking at the [Marchmallow docs](https://marshmallow.readthedocs.io/en/latest/examples.html#quotes-api-flask-sqlalchemy) it seems that they can validate whether the JSON passed to the API describes an application-specific Model instance, but they don't offer a function that takes valid JSON and creates an SQLA object. That is, they wind up iterating over the dictionary and assigning values to attributes. So, I changed the signature to take a dictionary (ideally, the one that results from `ItemSchema.load(json_data)`)

- Saw that you created a `create_item` method in the controller. Have you changed your mind about using Marshmallow? Should I be validating the dictionaries that you are passing?

- we should think about the use case for a client-facing /items/delete/<code>... Also, there's no reason that the DALItem.delete_item() method culdn't take a PK, which we know is indexed and an integer value...

jpcasey commented:
> * It's debatable whether we should have client-facing `add_item` and `add_store` endpoints
Yeah I'm not sure of a use case for them in the app, unless we decided to have some sort of admin web interface (if that makes anything easier).  I mainly just added them for testing.

> * Saw that you created a `create_item` method in the controller. Have you changed your mind about using Marshmallow? Should I be validating the dictionaries that you are passing?
I'll look at _de_serializing with Marshmallow later.  That method is just what was already in the add item endpoint.  I don't think you should need to do validation in the DAL.  That's best handled in the controller.

> * we should think about the use case for a client-facing `/items/delete/<code>`... Also, there's no reason that the `DALItem.delete_item()` method couldn't take a PK, which we know is indexed and an integer value
DELETE methods were again mainly just for testing.  Good point.  Anything the client won't be calling could still use a PK.
